File: C++/75. Sort Colors.cpp
#include <bits/stdc++.h>

using namespace std;

// Binary Search Method
void sortColors(vector<int> &nums)
{
    int cur = 0, left = 0, right = nums.size() - 1;

    while (cur <= right)
    {
        if (nums[cur] == 0)
            swap(nums[cur++], nums[left++]);
        else if (nums[cur] == 2)
            swap(nums[cur], nums[right--]);
        else
            cur++;
    }
}

// 2-pass
void sortColors(vector<int> &nums)
{
    int white = 0, blue = 0, n = nums.size();

    for (int x : nums)
    {
        if (x == 0)
            white++;
        else if (x == 1)
            blue++;
    }

    int i = 0;
    while (i < white)
        nums[i++] = 0;

    while (i < blue)
        nums[i++] = 1;

    while (i < n)
        nums[i++] = 2;
}

File: Rust/75. Sort Colors.rs
// simple solution
pub fn sort_colors(nums: &mut Vec<i32>) {
    let mut red = 0;
    let mut blue = 0;

    for i in 0..nums.len() {
        if nums[i] == 0 {
            red += 1;
        } else if nums[i] == 1 {
            blue += 1;
        }
    }

    let mut idx = 0;
    for _ in 0..red {
        nums[idx] = 0;
        idx += 1;
    }

    for _ in 0..blue {
        nums[idx] = 1;
        idx += 1;
    }

    for i in idx..nums.len() {
        nums[i] = 2;
    }
}

// binary search solution
pub fn sort_colors(nums: &mut Vec<i32>) {
    let mut low = 0;
    let mut mid = 0;
    let mut high = nums.len() - 1;

    while mid <= high {
        if nums[mid] == 0 {
            nums.swap(low, mid);
            low += 1;
            mid += 1;
        } else if nums[mid] == 1 {
            mid += 1;
        } else {
            nums.swap(high, mid);
            if high > 0 {
                high -= 1;
            } else {
                break;
            }
        }
    }
}
